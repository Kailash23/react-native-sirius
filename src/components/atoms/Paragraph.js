import * as React from 'react';
import HtmlView from 'react-native-htmlview';
import { Linking } from 'react-native';
import { withTheme } from 'react-native-paper';
import { customFonts } from '@styles/customFonts';
import { SIZES, PADDING_BW_ITEMS, getFontSize } from '@styles/sizes';
import InAppBrowser from 'react-native-inappbrowser-reborn';

function Paragraph(props) {
  let ParaFont = getFontSize(SIZES.PARA_FONTSIZE);
  let baseFont = getFontSize(SIZES.BASE);
  const styles = {
    container: {
      paddingHorizontal: SIZES.PaddingHorizontal
    },

    text: {
      fontFamily: customFonts.torstarTextRoman,
      fontSize: ParaFont,
      lineHeight: getFontSize(SIZES.LINE_HEIGHT)
    },

    em: {
      fontStyle: 'italic'
    },

    i: {
      fontStyle: 'italic'
    },

    u: {
      textDecorationLine: 'underline'
    },

    b: {
      fontWeight: '700'
    },

    strong: {
      fontWeight: '700'
    },

    h1: {
      fontWeight: '700',
      fontSize: ParaFont
    },

    h2: {
      fontWeight: '700',
      fontSize: ParaFont
    },

    h3: {
      fontWeight: '700',
      fontSize: ParaFont
    },

    a: {
      color: props.theme.colors.hyperLinks,
      textDecorationLine: props.theme.colors.hyperLinkStyle,
      // nested color style gets ignored on some android device, adding backgroundColor can solve the issue
      backgroundColor: 'transparent'
    },

    li: {
      flex: 1
    },

    bullet: {
      marginTop: 1,
      marginRight: 5,
      fontSize: ParaFont
    },

    sup: {
      fontSize: baseFont
    },

    p: {
      paddingBottom: PADDING_BW_ITEMS,
      fontSize: ParaFont
    }
  };
  const handleLinkPress = async url => {
    if (url.includes('mailto')) {
      Linking.openURL(url).catch(err => {
        console.log('Failed to open URL:', err.message);
      });
    } else {
      await InAppBrowser.open(url);
    }
  };

  return (
    <HtmlView
      value={props.data.text}
      stylesheet={styles}
      addLineBreaks={false}
      onLinkPress={url => handleLinkPress(url)}
      style={styles.container}
      textComponentProps={{
        style: [styles.text, { color: props.theme.colors.text }]
      }}
    />
  );
}

export default withTheme(Paragraph);
