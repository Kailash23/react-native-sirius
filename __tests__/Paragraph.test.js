import Paragraph from '@atoms/Paragraph';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';

configure({ adapter: new Adapter() });

describe('Testing <Paragraph/>', () => {
  test('Paragraph should render properly with the given html paragraph', () => {
    const wrapper = shallow(
      <Paragraph
        data={{
          text:
            '<p>Scotlynn Growers, a massive Norfolk County farming operation about two hours southwest of Toronto that describes itself as &#8220;North America&#8217;s Farm Stand,&#8221; is the site of one of Ontario&#8217;s largest recorded COVID-19 outbreaks. Some 199 workers there have tested positive for the virus.</p>'
        }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  test('Paragraph should render properly when Hyperlink is available in the paragraph ', () => {
    const wrapper = shallow(
      <Paragraph
        data={{
          text:
            '<p>In 1988, <a href="https://www.thestar.com/news/gta/2016/04/18/sius-creation-came-amidst-racially-charged-atmosphere.html">Lester Donaldson</a> was shot dead by a police officer in the Toronto rooming house where he lived. Donaldson, a 44-year-old Black man, was holding a small paring knife and had been diagnosed with schizophrenia.</p>'
        }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  test('Paragraph should render nothing if paragraph is empty ', () => {
    const wrapper = shallow(
      <Paragraph
        data={{
          text: ''
        }}
      />
    );
    expect(wrapper).toEqual({});
  });
});
