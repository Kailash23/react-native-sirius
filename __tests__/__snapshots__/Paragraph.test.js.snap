// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Paragraph/> Paragraph should render properly when Hyperlink is available in the paragraph  1`] = `
<ThemedComponent
  _reactThemeProviderForwardedRef={null}
  data={
    Object {
      "text": "<p>In 1988, <a href=\\"https://www.thestar.com/news/gta/2016/04/18/sius-creation-came-amidst-racially-charged-atmosphere.html\\">Lester Donaldson</a> was shot dead by a police officer in the Toronto rooming house where he lived. Donaldson, a 44-year-old Black man, was holding a small paring knife and had been diagnosed with schizophrenia.</p>",
    }
  }
/>
`;

exports[`Testing <Paragraph/> Paragraph should render properly with the given html paragraph 1`] = `
<ThemedComponent
  _reactThemeProviderForwardedRef={null}
  data={
    Object {
      "text": "<p>Scotlynn Growers, a massive Norfolk County farming operation about two hours southwest of Toronto that describes itself as &#8220;North America&#8217;s Farm Stand,&#8221; is the site of one of Ontario&#8217;s largest recorded COVID-19 outbreaks. Some 199 workers there have tested positive for the virus.</p>",
    }
  }
/>
`;
